lib/sparesults/src/json.rs

```diff
@@ -224,520 +220,882 @@ fn write_json_term<'a>(output: &mut Vec<JsonEvent<'a>>, term: TermRef<'a>) {
     }
 }
 
-pub enum JsonQueryResultsReader<R: Read> {
+pub enum FromReadJsonQueryResultsReader<R: Read> {
     Solutions {
         variables: Vec<Variable>,
-        solutions: JsonSolutionsReader<R>,
+        solutions: FromReadJsonSolutionsReader<R>,
     },
     Boolean(bool),
 }
 
-impl<R: Read> JsonQueryResultsReader<R> {
+impl<R: Read> FromReadJsonQueryResultsReader<R> {
     pub fn read(read: R) -> Result<Self, QueryResultsParseError> {
         let mut reader = FromReadJsonReader::new(read);
-        let mut variables = None;
-        let mut buffered_bindings: Option<Vec<_>> = None;
-        let mut output_iter = None;
-
-        if reader.read_next_event()? != JsonEvent::StartObject {
-            return Err(
-                QueryResultsSyntaxError::msg("SPARQL JSON results should be an object").into(),
-            );
+        let mut inner = JsonInnerReader::new();
+        loop {
+            if let Some(result) = inner.read_event(reader.read_next_event()?)? {
+                return match result {
+                    JsonInnerQueryResults::Solutions {
+                        variables,
+                        solutions,
+                    } => Ok(Self::Solutions {
+                        variables,
+                        solutions: FromReadJsonSolutionsReader {
+                            inner: solutions,
+                            reader,
+                        },
+                    }),
+                    JsonInnerQueryResults::Boolean(value) => Ok(Self::Boolean(value)),
+                };
+            }
         }
+    }
+}
+
+pub struct FromReadJsonSolutionsReader<R: Read> {
+    inner: JsonInnerSolutions,
+    reader: FromReadJsonReader<R>,
+}
+
+impl<R: Read> FromReadJsonSolutionsReader<R> {
+    pub fn read_next(&mut self) -> Result<Option<Vec<Option<Term>>>, QueryResultsParseError> {
+        match &mut self.inner {
+            JsonInnerSolutions::Reader(reader) => loop {
+                let event = self.reader.read_next_event()?;
+                if event == JsonEvent::Eof {
+                    return Ok(None);
+                }
+                if let Some(result) = reader.read_event(event)? {
+                    return Ok(Some(result));
+                }
+            },
+            JsonInnerSolutions::Iterator(iter) => iter.next(),
+        }
+    }
+}
 
+#[cfg(feature = "async-tokio")]
+pub enum FromTokioAsyncReadJsonQueryResultsReader<R: AsyncRead + Unpin> {
```

Comment:
ouch, the size of this struct name is nearly unreadable :(  Any way to make it shorter?

Reply:
Yes, that's awful. `TokioAsyncRead` is more or less mandatory because we want to state this is the wrapper for an `AsyncRead` from Tokio (and not the possible future ones for `futures::io::AsyncRead` and the possible `std::io::AsyncRead`...), `Json` to distinguish with `Xml` and `Reader` to state what the struct does.

I can maybe drop `QueryResults` because this is an internal type but I won't skip the `QueryResults` part in the public API to avoid name clash with the RDF parsers so it's not a great idea.

What I can easily do is drop the `From` (and `To` in the serializers). What do you think?

Reply:
Naming is hard!  (tm)  No idea tbh